table.cc

```diff
@@ -2476,15 +2476,29 @@ table::make_reader_excluding_sstable(schema_ptr s,
     return make_combined_reader(s, std::move(readers), fwd, fwd_mr);
 }
 
-void table::move_sstable_from_staging_in_thread(sstables::shared_sstable sst) {
-    try {
-        sst->move_to_new_dir(dir(), sst->generation()).get();
-    } catch (...) {
-        tlogger.warn("Failed to move sstable {} from staging: {}", sst->get_filename(), std::current_exception());
-        return;
-    }
-    _sstables_staging.erase(sst->generation());
-    _compaction_strategy.get_backlog_tracker().add_sstable(sst);
+future<> table::move_sstables_from_staging(std::vector<sstables::shared_sstable> sstables) {
+    return with_semaphore(_sstable_deletion_sem, 1, [this, sstables = std::move(sstables)] {
+        return do_with(std::set<sstring>({dir()}), [this, sstables = std::move(sstables)] (auto& dirs_to_sync) {
+            return parallel_for_each(std::move(sstables), [this, &dirs_to_sync] (sstables::shared_sstable sst) {
```

Comment:
Have you considered how high can parallelism get here? My question is: can there be, say, 100k staging tables queued for moving from staging at a time, e.g. because something keeps failing and the queue grows and grows? Because if that's plausible, then we have a problem, especially that `move_to_new_dir` also leverages `parallel_for_each` in multiple places, and we could end up with millions of continuations getting created here each time this function runs.

I hope the scenario above is impossible and I just missed some details, but I'd like a confirmation :)

Reply:
It's unlikely, but with a configuration of thousands of tables having views we may end up with thousands of sstables to move.

Reply:
Since we're now syncing the directories at the end we can just move the sstables one at a time. I don't think that parallelism will buy us much in this case.

Reply:
It does get us something. If the server is busy each syscall operation ends up taking half a millisecond, and a move may be several such non-parallel operations. So moving tens of thousands of sstables can take tens of seconds.

Ok, maybe this isn't so important.

Reply:
also. please note we do the move operations on all the components of each sstable in parallel in `sstable::create_links` and then in `sstable::move_to_new_dir`